bitcoin/src/psbt/mod.rs

```diff
@@ -844,13 +849,50 @@ impl GetKey for $map<PublicKey, PrivateKey> {
     ) -> Result<Option<PrivateKey>, Self::Error> {
         match key_request {
             KeyRequest::Pubkey(pk) => Ok(self.get(&pk).cloned()),
+            KeyRequest::XOnlyPubkey(xonly) => {
+                let pubkey_even = PublicKey::new(xonly.public_key(secp256k1::Parity::Even));
+                let key = self.get(&pubkey_even).cloned();
+                
+                if key.is_some() {
+                    return Ok(key);
+                }
+                
+                let pubkey_odd = PublicKey::new(xonly.public_key(secp256k1::Parity::Odd));
+                Ok(self.get(&pubkey_odd).cloned())
+            },
+            KeyRequest::Bip32(_) => Err(GetKeyError::NotSupported),
+        }
+    }
+}}}
+impl_get_key_for_pubkey_map!(BTreeMap);
+#[cfg(feature = "std")]
+impl_get_key_for_pubkey_map!(HashMap);
+
+#[rustfmt::skip]
+macro_rules! impl_get_key_for_xonly_map {
+    ($map:ident) => {
+
+impl GetKey for $map<XOnlyPublicKey, PrivateKey> {
+    type Error = GetKeyError;
+
+    fn get_key<C: Signing>(
+        &self,
+        key_request: &KeyRequest,
+        _: &Secp256k1<C>,
+    ) -> Result<Option<PrivateKey>, Self::Error> {
+        match key_request {
+            KeyRequest::XOnlyPubkey(xonly) => Ok(self.get(xonly).cloned()),
+            KeyRequest::Pubkey(pk) => {
+                let xonly = XOnlyPublicKey::from(pk.inner);
+                Ok(self.get(&xonly).cloned())
```

Comment:
There's one slight issue: after retrieving the key we need to check the parity of `pk` and negate the private key if it's `Odd`.